lib/helpers/wp-cli-db/class-wp-cli-db.php

```diff
@@ -30,40 +31,83 @@ public function early_init() {
 	}
 
 	/**
-	 * Ensure the command is allowed for the current Config.
+	 * Ensure the command or query is allowed for the current Config.
 	 *
-	 * @throws Exception if the command is not allowed.
+	 * @param array $command
+	 * @return WP_Error|void
 	 */
-	public function validate_subcommand( array $command ): void {
+	public function validate_subcommand( array $command ) {
 		$subcommand = $command[1] ?? '';
 
-		if ( 'query' !== $subcommand ) {
-			throw new Exception( 'ERROR: Only the `wp db query` subcommand is permitted for this site.' );
+		$allowed_subcommands = [
+			'query',
+			'prefix',
+			'columns',
+			'size',
+		];
+
+		if ( ! in_array( $subcommand, $allowed_subcommands, true ) ) {
+			return new \WP_Error( 'db-cli-disallowed-subcmd', "The `wp db $subcommand` subcommand is not permitted for this site." );
+		}
+
+		if ( 'query' === $subcommand && 2 === count( $command ) ) {
+			// Doing `wp db query` without a DB query is the equivalent of doing `wp db cli`
+			return new \WP_Error( 'db-cli-missing-query', 'Please provide the database query as a part of the command.' );
 		}
 
-		if ( 2 === count( $command ) ) {
-			throw new Exception( 'ERROR: Please provide the database query as a part of the command.' );
+		$query      = $command[2];
+		$validation = $this->validate_query( $query );
+		if ( is_wp_error( $validation ) ) {
+			WP_CLI::error( $validation->get_error_message() );
+		}
+	}
+
+	/**
+	 * Ensure the query is allowed.
+	 *
+	 * @param string $query
+	 * @return WP_Error|void
+	 */
+	public function validate_query( string $query ) {
```

Comment:
I'd modify the signture to return bool|WP_Error and add `return true` at the end. Which woud signify that the query is valid.

Reply:
I wonder if that would be confusing because someone may expect it to return false on invalidation 🤔